e2e2/test/images/bert-training/train.py

```diff
@@ -0,0 +1,139 @@
+    )
+
+
+def mask_tokens(inputs, tokenizer, mlm_probability):
+    labels = inputs.clone()
+    probability_matrix = torch.full(labels.shape, mlm_probability)
+    special_tokens_mask = [
+        tokenizer.get_special_tokens_mask(val, already_has_special_tokens=True)
+        for val in labels.tolist()
+    ]
+    probability_matrix.masked_fill_(
+        torch.tensor(special_tokens_mask, dtype=torch.bool), value=0.0
+    )
+    masked_indices = torch.bernoulli(probability_matrix).bool()
+    labels[~masked_indices] = -100  # We only compute loss on masked tokens
+
+    inputs[masked_indices] = tokenizer.convert_tokens_to_ids(tokenizer.mask_token)
+
+    return inputs, labels
+
+
+def setup(rank, world_size, local_rank):
+    master_addr = os.environ["MASTER_ADDR"]
+    master_port = os.environ["MASTER_PORT"]
+    dist.init_process_group(
+        "nccl",
+        init_method=f"tcp://{master_addr}:{master_port}",
+        rank=rank,
+        world_size=world_size,
+    )
+    torch.cuda.set_device(local_rank)
+    print(f"Process {rank} initialized, using GPU {local_rank}")
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def train_bert(rank, world_size, local_rank, model, tokenizer):
+    setup(rank, world_size, local_rank)
+
+    model = model.to(local_rank)
+    ddp_model = DDP(model, device_ids=[local_rank])
+
+    dataset = create_dummy_data(tokenizer)
+    train_dataloader = DataLoader(dataset, batch_size=8)
+
+    optimizer = torch.optim.AdamW(ddp_model.parameters(), lr=0.001)
+    criterion = torch.nn.CrossEntropyLoss()
+
+    start_time = time.time()
+
+    for epoch in range(1):  # Short run for testing
+        ddp_model.train()
+        for batch in train_dataloader:
+            optimizer.zero_grad()
+            inputs, masks, labels, next_sentence_labels = batch
+            inputs, masks, labels, next_sentence_labels = (
+                inputs.to(local_rank),
+                masks.to(local_rank),
+                labels.to(local_rank),
+                next_sentence_labels.to(local_rank),
+            )
+            outputs = ddp_model(
+                input_ids=inputs,
+                attention_mask=masks,
+                labels=labels,
+                next_sentence_label=next_sentence_labels,
+            )
+            loss = outputs.loss
+            loss.backward()
+            optimizer.step()
+
+    end_time = time.time()
+    training_time = end_time - start_time
+    throughput = len(dataset) / training_time
+
+    print(f"Process {rank} - Training time: {training_time:.2f} seconds")
+    print(f"Process {rank} - Throughput: {throughput:.2f} samples/second")
+
+    cleanup()
+
+
+def main():
+    rank = int(os.environ["OMPI_COMM_WORLD_RANK"])
+    world_size = int(os.environ["OMPI_COMM_WORLD_SIZE"])
+
+    # TODO: Consider parameterizing for nodes of any GPU count
+    num_gpus_per_node = 8  # Adjust this based on your setup
```

Comment:
we need to do this now right, in case we want to call a job with a different instance type entirely? i guess the testing instances have been the same capacity so far

Reply:
Yeah, at the moment all instance types requested have 8 GPUs. I don't really ever see us running the training test on any instance type with less than 8 GPUs, unless EC2 suddenly changes their pattern for instance configurations.

Reply:
It would future proof, but also add a touch more complexity and another point of failure if we have to have it configured at runtime.

Reply:
@ndbaker1 I'd actually probably advocate I take out the TODO comment, and just leave it hard coded. Thoughts?

Reply:
was thinking we could just add an
```Dockerfile
ENV GPUS_PER_NODE=8
```
in the Dockerfile and it wouldn't complicate much right?

Reply:
Yeah that works. I'm thinking further down stream (upstream..?) for parameterizitation of the manifest, and collection of the number of GPUs on a node by our Go test.

Reply:
So would really need to be something like

`ARG GPUS_PER_NODE=8`

unless I'm misunderstanding something.

Reply:
@ndbaker1 Made the change, verified with local testing